src/Solvers/multigrid_solver.jl

```diff
@@ -0,0 +1,144 @@
+using Oceananigans.Architectures: architecture
+using Oceananigans.Grids: interior_parent_indices
+using Statistics: norm, dot
+using LinearAlgebra
+using AlgebraicMultigrid: _solve!, init, RugeStubenAMG
+using Oceananigans.Operators: volume, Δyᶠᶜᵃ, Δyᶜᶠᵃ, Δyᶜᶜᵃ, Δxᶠᶜᵃ, Δxᶜᶠᵃ, Δxᶜᶜᵃ, Δyᵃᶜᵃ, Δxᶜᵃᵃ, Δzᵃᵃᶠ, Δzᵃᵃᶜ, ∇²ᶜᶜᶜ
+
+import Oceananigans.Architectures: architecture
+
+mutable struct MultigridSolver{A, G, L, T, M}
+               architecture :: A
+                       grid :: G
+            linear_operator :: L
+                  tolerance :: T
+         maximum_iterations :: Int
+              amg_algorithm :: M
+end
+
+architecture(solver::MultigridSolver) = solver.architecture
+    
+"""
+    MultigridSolver(grid,
+                    linear_operation!, 
+                    args...;
+                    maximum_iterations = 100,
+                    tolerance = 1e-13,
+                    amg_algorithm = RugeStubenAMG(),
+                    )
+
+Returns a MultigridSolver that solves the linear equation
+``A x = b`` using a multigrid method where `A * x` is
+determined by `linear_operation!`
+
+`linear_operation!` is a function with signature `linear_operation!(Ax, x, args...)` 
+that calculates `A*x` and stores the result in `Ax` for a "candidate solution `x`.
+
+
+The solver is used by calling
+
+```
+solve!(x, solver::MultigridSolver, b; kwargs...)
+```
+
+for `solver`, right-hand side `b`, solution `x`, and optional keyword arguments `kwargs...`.
+
+Arguments
+=========
+
+* `maximum_iterations`: Maximum number of iterations the solver may perform before exiting.
+
+* `tolerance`: Tolerance for convergence of the algorithm. The algorithm quits when
+               `norm(A * x - b) < tolerance`.
+"""
+function MultigridSolver(grid,
+                        linear_operation!, 
+                        args...;
+                        maximum_iterations = 100, #prod(size(template_field)),
+                        tolerance = 1e-13, #sqrt(eps(eltype(template_field.grid))),
+                        amg_algorithm = RugeStubenAMG(),
+                        )
+
+    arch = architecture(grid)
+
+    A = create_matrix(grid, linear_operation!, args...)
+
+    return MultigridSolver(arch,
+                           grid,
+                           A,
+                           tolerance,
+                           maximum_iterations,
+                           amg_algorithm
+                           )
+end
+
+# TODO make inplace create_matrix!
+
+# For free surface without Δt
+function create_matrix(grid, linear_operator!, ::Any , ::Any, ::Any, ::Nothing)
+    Nx, Ny, Nz = grid.Nx, grid.Ny, grid.Nz
+    return spzeros(eltype(grid), Nx*Ny, Nx*Ny)
+end
+
```

Comment:
```suggestion
```

Reply:
Why should we delete this? It's the `Δt` is not initialised case, is there a nicer way to do this?

Reply:
Yes, but I thought that we figured out that `Δt = nothing` wasn't working and we decided to initialize with `Δt = -1.0`. So this is never used, right?